tests/test_advection.py

```diff
@@ -1,167 +1,224 @@
-
-    # interpolate initial conditions
-    f.interpolate(fexpr)
-    f_end.interpolate(f_end_expr)
-    state.initialise([('u', u0), ('f', f)])
-
-    if spatial_opts is None:
-        fequation = AdvectionEquation(state, f.function_space(), ibp=ibp, equation_form=equation_form)
-    elif "supg_params" in spatial_opts:
-        fequation = SUPGAdvection(state, f.function_space(), ibp=ibp, equation_form=equation_form, supg_params=spatial_opts["supg_params"])
-    elif "embedded_dg" in spatial_opts:
-        if spatial_opts["embedded_dg"]["space"] == "Broken":
-            fequation = EmbeddedDGAdvection(state, f.function_space(), ibp=ibp, equation_form=equation_form)
-        elif spatial_opts["embedded_dg"]["space"] == "DG":
-            fequation = EmbeddedDGAdvection(state, f.function_space(), ibp=ibp, equation_form=equation_form, Vdg=space)
 
+def setup_model(state, physical_domain, timestepping, field_eqns, time_discretisation):
+
+    advected_fields = []
     if time_discretisation == "ssprk":
-        f_advection = SSPRK3(state, f, fequation)
+        for fname, equation in field_eqns.items():
+            advected_fields.append((fname, SSPRK3(state.fields(fname),
+                                                  timestepping.dt,
+                                                  equation)))
     elif time_discretisation == "implicit_midpoint":
-        f_advection = ThetaMethod(state, f, fequation)
+        for fname, equation in field_eqns.items():
+            advected_fields.append((fname, ThetaMethod(state.fields(fname),
+                                                       timestepping.dt,
+                                                       equation)))
 
-    advected_fields = [("f", f_advection)]
-    timestepper = AdvectionTimestepper(state, advected_fields)
+    return AdvectionDiffusionModel(state, physical_domain,
+                                   timestepping=timestepping,
+                                   advected_fields=advected_fields)
 
-    return timestepper, tmax, f_end
 
+def run(model, tmax, fieldlist):
 
-def run(dirname, geometry, time_discretisation, ibp, equation_form, vector, spatial_opts=None):
+    timestepper = AdvectionTimestepper(model)
+    f_dict = timestepper.run(0, tmax, x_end=fieldlist)
+    return f_dict
 
-    timestepper, tmax, f_end = setup_advection(dirname, geometry, time_discretisation, ibp, equation_form, vector, spatial_opts=spatial_opts)
 
-    f_dict = timestepper.run(0, tmax, x_end=["f"])
-    f = f_dict["f"]
+def check_errors(f_dict, f_end, field_eqns, tol):
+    f_err = Function(f_end.function_space())
+    errors = {}
+    for name, field in f_dict.items():
+        f_err.assign(f_end - field)
+        errors[name] = max(abs(f_err.dat.data.min()), abs(f_err.dat.data.max()))
+    for name, err in errors.items():
+        if err > tol:
+            print("Test fails for field %s with options ibp = %s, continuity = %s" % (name, field_eqns[name].ibp, field_eqns[name].continuity))
```

Comment:
@wence- How about here if field_eqns isn't a dict? Is it safe to use `dict(field_eqns)` [note: I do this in `model.py` - see e.g. line 95] or is there a better way to search a list of tuples?

I am trying to print out some useful information because there are multiple asserts here and I want to know exactly which one failed. Maybe this is the wrong way to go about it - these used to be separate tests but they take a fairly long time and I wondered if testing all the advection options by using different fields within one simulation might speed things up by avoiding repeating the model setup.